### PR TITLE
Cachito integration plugin

### PR DESCRIPTION
**Pull request by lcarva**
Maintainers will complete the following section:
- [x] Commit messages are descriptive enough
- [x] "Signed-off-by:" line is present in each commit
- [x] Code coverage from testing does not decrease and new code is covered
- [x] JSON/YAML configuration changes are updated in the relevant schema
- [x] Changes to metadata also update the documentation for the metadata
- [x] Pull request includes link to an osbs-docs PR for user documentation updates
- [x] New feature can be disabled from a configuration file

### CODE REVIEW

File: atomic_reactor/plugins/pre_reactor_config.py [hunk 1]
```diff
@@ -131,6 +131,27 @@ def get_smtp_session(workflow, fallback):
     return smtplib.SMTP(config['host'])
 
 
+def get_cachito(workflow):
+    return get_value(workflow, 'cachito', NO_FALLBACK)
+
+
+def get_cachito_session(workflow):
+    config = get_cachito(workflow)
+    from atomic_reactor.cachito_util import CachitoAPI
```
Comment (MartinBasti) [inline atomic_reactor/plugins/pre_reactor_config.py]: Why is this import here in function? Cyclic imports?

File: atomic_reactor/plugins/pre_reactor_config.py [hunk 2]
```diff
@@ -131,6 +131,27 @@ def get_smtp_session(workflow, fallback):
...
+
+    api_kwargs = {'insecure': config.get('insecure', False)}
+
+    ssl_certs_dir = config['auth'].get('ssl_certs_dir')
+    if ssl_certs_dir:
+        cert_path = os.path.join(ssl_certs_dir, 'cert')
+        if os.path.exists(cert_path):
+            api_kwargs['cert'] = cert_path
+        else:
+            raise KeyError("Cachito ssl_certs_dir doesn't exist")
```
Comment (MartinBasti) [inline atomic_reactor/plugins/pre_reactor_config.py]: IMHO it should be `ValueError`

File: atomic_reactor/schemas/container.json [hunk 3]
```diff
@@ -156,6 +156,51 @@
       "description": "build-step plugin to be used",
       "enum": ["docker_api", "imagebuilder", "buildah_bud"]
     },
+    "remote_source": {
```
Comment (MartinBasti) [inline atomic_reactor/schemas/container.json]: Wouldn't be better to having it in this format for future extension:

```
remote_sources: [
    {
      "repo": "git.example.com/proj-go",
      "ref": "b55c00f45ec3dfee0c766cea3d395d6e21cc2e5a",
      "pkg_manager": "gomod",
      "flags": [...cachito flags...],
   }, {
      "repo": "another.git.com/pyproject",
      "ref": "651abc41170e52866148cc0c99c3671a2ace7002",
      "pkg_manager": "pip",
      "flags": [...cachito flags...],
   }
]
```

File: atomic_reactor/plugins/pre_reactor_config.py (hunk 2 above)
Comment (MartinBasti) [reply atomic_reactor/plugins/pre_reactor_config.py]: Or `RuntimeError`

File: atomic_reactor/plugins/pre_reactor_config.py (hunk 1 above)
Comment (lcarva) [reply atomic_reactor/plugins/pre_reactor_config.py]: I followed the pattern in `get_odcs_session`. IIRC, this is so we can split up the plugin into subpackages. If that's not needed, I can move it to the top.

File: atomic_reactor/plugins/pre_reactor_config.py (hunk 2 above)
Comment (lcarva) [reply atomic_reactor/plugins/pre_reactor_config.py]: It's `KeyError` because that's what `get_value` raises if a config for `cachito` doesn't exist at all. I suppose `get_odcs_session` does this so it's easier to use by only catching a single exception.

File: atomic_reactor/schemas/container.json (hunk 3 above)
Comment (lcarva) [reply atomic_reactor/schemas/container.json]: During the design phase, we decided to only support a single remote source. But a single remote source could have multiple package managers. This is inline with upstream projects.

Having multiple remote sources is challenging because if you have, for example, two goloang apps, the value for GOCACHE/GOPATCH is ambiguous. 

Given the complexities in implementing this, we decided that this could be added later on if there was need.

Comment (lcarva) [reply atomic_reactor/schemas/container.json]: During the design phase we decided to only support a single remote source, but allow multiple package managers for it. 

Supporting multiple remote sources can be challenging. For instance, if there are two remote sources using gomod, it's unclear what the value of GOCACHE/GOPATH should be set to. Should it point to the cache of the first remote source, or the second?

Given the complexities that can arise from this scenario, we decided to only support a single remote source until a real use case arises to makes us reconsider.

File: atomic_reactor/plugins/pre_reactor_config.py (hunk 1 above)
Comment (MartinBasti) [reply atomic_reactor/plugins/pre_reactor_config.py]: Please put it on top of the module

File: atomic_reactor/schemas/container.json (hunk 3 above)
Comment (MartinBasti) [reply atomic_reactor/schemas/container.json]: ok

File: atomic_reactor/plugins/pre_reactor_config.py (hunk 1 above)
Comment (lcarva) [reply atomic_reactor/plugins/pre_reactor_config.py]: Done!

File: atomic_reactor/plugins/pre_reactor_config.py (hunk 2 above)
Comment (lcarva) [reply atomic_reactor/plugins/pre_reactor_config.py]: You're right though. `KeyError` doesn't make any sense. I changed it to `RuntimeError`.

**Comment by lcarva** [comment]
Added issue ID to each commit. I think I addressed all the comments.

**Comment by MartinBasti** [comment]
no release notes, when feature is done a beautiful release notes will be added
